components/merkledb/src/db.rs

```diff
@@ -642,26 +726,92 @@ impl<'a> IndexAccess for &'a Fork {
     }
 }
 
-impl IndexAccess for Rc<Fork> {
-    type Changes = ChangesRef<'static>;
+impl RawAccess for Rc<Fork> {
+    type Changes = ChangesMut<'static>;
 
     fn snapshot(&self) -> &dyn Snapshot {
         &self.patch
     }
 
     fn changes(&self, address: &IndexAddress) -> Self::Changes {
         let changes = self.working_patch.take_view_changes(address);
-        ChangesRef {
+        ChangesMut {
             changes,
             key: address.clone(),
             parent: WorkingPatchRef::Owned(Self::clone(self)),
         }
     }
 }
 
-impl AsRef<dyn Snapshot> for Fork {
-    fn as_ref(&self) -> &dyn Snapshot {
-        &self.patch
+/// Readonly wrapper for a `Fork`.
+///
+/// This wrapper allows to read from index state from the fork
+/// in a type-safe manner (it is impossible to accidentally modify data in the index), and
+/// without encountering runtime errors when attempting to concurrently get the same index
+/// more than once.
+///
+/// Since the wrapper borrows the `Fork` immutably, it is still possible to access indexes
+/// in the fork directly. In this scenario, the caller should be careful that `ReadonlyFork`
```

Comment:
Well, it could be possible for `ReadonlyFork` to wrap `&mut Fork` instead of `&Fork`, but that would defeat the purpose of `ReadonlyFork`. In Exonum, we *want* to have concurrent access to different parts of the database via the same `Fork`: full access to the data of the executing service, and readonly access to the core / dispatcher and other services.